bpf/lib/nat.h

```diff
@@ -538,21 +543,80 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
 		tuple.dport = l4hdr.dport;
 		tuple.sport = l4hdr.sport;
 		break;
+#ifndef DISABLE_SNAT_ICMP_SUPPORT
 	case IPPROTO_ICMP:
 		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
 			return DROP_INVALID;
-		if (icmphdr.type != ICMP_ECHO &&
-		    icmphdr.type != ICMP_ECHOREPLY)
-			return DROP_NAT_UNSUPP_PROTO;
-		if (icmphdr.type == ICMP_ECHO) {
+
+		switch (icmphdr.type) {
+		case ICMP_ECHO:
 			tuple.dport = 0;
 			tuple.sport = icmphdr.un.echo.id;
-		} else {
+			break;
+		case ICMP_ECHOREPLY:
 			tuple.dport = icmphdr.un.echo.id;
 			tuple.sport = 0;
 			icmp_echoreply = true;
+			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED:
+				/* According to the RFC 5508, any networking
+				 * equipment that is responding with an ICMP Error
+				 * packet should embed the original packet in its
+				 * response.
+				 */
+
+				off += sizeof(icmphdr);
+				if (ctx_load_bytes(ctx, off, &iphdr,
+						   sizeof(iphdr)) < 0)
+					return DROP_INVALID;
+
+				/* From the embedded IP headers we should be able
+				 * to determine corresponding protocol, IP src/dst
+				 * of the packet sent to resolve the NAT session.
+				 */
+				tuple.nexthdr = iphdr.protocol;
+				tuple.saddr = iphdr.daddr;
+				tuple.daddr = iphdr.saddr;
```

Comment:
@brb The change is done here.

```
From original IP hdr:
--
tuple.daddr is 192.168.49.1   # HOST
tuple.saddr is 192.168.49.2   # ROUTER

Embeded packet IP hdr:
--
iphdr.saddr is 192.168.49.1 # HOST
iphdr.daddr is 10.1.0.100     # DST

Changes
--
tuple.saddr is 10.1.0.100     # DST
tuple.daddr is 192.168.49.1 # HOST

```

1. Here [0] we can see that we create a nat entry with `to_daddr` pointing to the  POD.
2. Here [1] we can see that we create a rtuple entry with saddr pointing to DST and daddr pointing to HOST



[0] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L218
[1] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L229

You probably know all of that but since i'm not sure to understand your whole question I'm basically trying to share with you my understanding.

Reply:
Could you please attach a pcap with the ICMP NEED FRAG. I'd help me with reviewing the changes.

Reply:
Sure give me a moment to build env and share this with you. Thank you.

Reply:
@brb Morning,

Here are the pcap from pod and bridge POV.

https://drive.google.com/drive/folders/1BFrLFkXUlyy_sUZ6ryElcI9Il2p-Zhnf?usp=sharing

Let me know if you need anything else. Thanks

Reply:
Hello Martynas, 

Thank you to have looked at it so fast :-)

My understand is that. During egress we store a rtuple DST IP/HOST IP with a nat entry pointing to the POD IP. during ingress we enter in the function `snat_v4_rewrite_ingress` [0] to rewite the packet destination.

[0] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L422

Reply:
There is effectively something that is not clear, in the embedded packet the source IP is the router so it seems that we endup with a tuple ROUTER IP/POD IP, the snat does not have such entry in its table...

Reply:
```
Pod-X      : 10.244.0.236
Cilium-Nat : 192.168.49.2
Router-Y   : 192.168.49.1
Host-Y     : 10.1.0.100

                                         
                                  +------------+
                                  | Pod-X      |
                                  +------------+
            Cilium Host                  |
            --------------------+--------+-------------------
                                |  | ICMP Packet Pod-X to Host-y
                                |  |   10.244.0.256/10.1.0.100
                                |  v
                          +--------------+
                          | Cilium-Nat   |
                          +--------------+
                                |  | ICMP Packet Cilium-Nat to Host-y
                                |  |   192.168.49.2/10.1.0.100 
    World                       |  v
   ----------------+------------+----------------
                   |  ^
                   |  | ICMP Error Router-Y to Cilium-Nat
                   |  |   192.168.49.1/192.168.49.2
                   |  |      with in Payload ICMP 192.168.49.2/10.1.0.100
            +-------------+
            | Router-Y    |
            +-------------+
                   |
   ----------------+-------+--------
                           |
                         Host-Y
```

Hello Martynas, I have provided some more details about the setup and scenario. I hope that will help. Let me know if you need any more information. Thanks.

Reply:
Nothing is mentioning that the payload should be updated. We should even not drop the packet if the embedded one has an invalid checksum [0]. At least from my understanding.

We are doing end-to-end tests with real use-case using IPSec. I will let you know the result when it's done.

https://www.rfc-editor.org/rfc/rfc5508.html#section-4.1

Reply:
Could you check what netfilter's SNAT does? You could disable Cilium's BPF masq by setting `--enable-bpf-masquerade=false`.

Reply:
I should have think about that before :-) thank you I will do that and let you know.

Reply:
Hello Martynas, It seems that the source IP of the  embbeded packet it also translated, meaning the checksum of the ip header is translated as-well. It's quite a big thing to implement this feature. I will add this and get back to you shortly.

  https://drive.google.com/file/d/1ZpGLNOMG6P4-UbOmDS2mdRv2manO-Pun/view

Reply:
Hey @sahid , I think I am still confused by the output seen in https://github.com/cilium/cilium/pull/18414/files#r827040230. Your change is supposed to rev-SNAT the embedded packet in the ICMP FRAG NEEDED. But in the output we can see that the ICMP FRAG NEEDED is rev-SNAT-ed instead :thinking: 

Another thing is that the required changes might explode the complexity. I'm taking to my this week's TODO to check whether the refactoring snat_v{4,6}_process => do_snat_v{4,6}/do_rev_snat_v{4,6} is feasible. This could help with your PR.

Reply:
> Hey @sahid , I think I am still confused by the output seen in https://github.com/cilium/cilium/pull/18414/files#r827040230. Your change is supposed to rev-SNAT the embedded packet in the ICMP FRAG NEEDED. But in the output we can see that the ICMP FRAG NEEDED is rev-SNAT-ed instead thinking

Hum... no no the change is well supposed to rev-SNAT the ICMP FRAG NEEDED. I think we are good for this part. Are you considering that we should not?

The missing part is to replace the source IP of the embedded packet from HOST to POD.

> Another thing is that the required changes might explode the complexity. I'm taking to my this week's TODO to check whether the refactoring snat_v{4,6}_process => do_snat_v{4,6}/do_rev_snat_v{4,6} is feasible. This could help with your PR.

Yes, I'm also a bit scary for that as we will have to rewrite two new fields for the embedded packet, the source IP and checksum.

Oh I actually wanted to work on it after this change, but if you have desires for that I will let you take the lead :-)

Reply:
> think we are good for this part. Are you considering that we should not?

Re-read the changes. I was mistakenly thinking, that inside the FRAG NEEDED branch you were updating `off`. All good.

> I wanted to work on it after this change, but if you have desires for that I will let you take the lead :-)

Sure, you can work on it. I just want to validate whether the proposed refactoring makes sense.

Reply:
Hello @brb the nat part for the embedded packet is ready, ipv4 only. When you have moment to make review and share your thinking., I understand the patch is quite big now and may be painful to review. Thanks in advance.

Reply:
Hum I just noticed your comment. I have shared the iptables example demonstrating that the embedded packet get NATed as-well. Also previously there was tests to demonstrate that the code was doing what is expected. But As mentioned in a different threads I had to remove all of them because Cilium proposed a new test framework.

All good now I guess. I have re-implemented tests using the new test framework. Unfortunately for IPv6 there are some issues. I hope that will not be a blocker. Thanks.